File: Polymorphism/PrototypeDemo.py
class Point:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y


class Shape:
    def __init__(self, start: Point, end: Point, color: int = 0, thickness: int = 1, fill_style: int = 1):
        self.start = start
        self.end = end
        self.color = color
        self.thickness = thickness
        self.fill_style = fill_style


class Line(Shape):
    def __init__(self, start: Point, end: Point, color: int = 0, thickness: int = 1, fill_style: int = 1):
        super().__init__(start, end, color, thickness, fill_style)

    def draw(self):
        print(
            f"Drawing Line from {self.start.x}:{self.start.y} to {self.end.x}:{self.end.y}")


class Rectangle(Shape):
    def __init__(self, start: Point, end: Point, color: int = 0, thickness: int = 1, fill_style: int = 1):
      super().__init__(start, end, color, thickness, fill_style)

    def draw(self):
        print(
            f"Drawing Rectangle from {self.start.x}:{self.start.y} to {self.end.x}:{self.end.y}")


shape = Line(Point(0, 0), Point(10, 10))
shape.draw()

File: Polymorphism/PrototypeDemo.js
class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}

class Shape {
  constructor(start, end, color=0, thickness=1, fillStyle=1) {
    this.start = start;
    this.end = end;
    this.color = color;
    this.thickness = thickness;
    this.fillStyle = fillStyle;
  }

  draw() {}

  clone(){}
}

class Line extends Shape {
  draw() {
    console.log(`Drawing Line from ${this.start.x}:${this.start.y} to ${this.end.x}:${this.end.y}`);
  }

  clone(){
    return this
  }
}

class Rectangle extends Shape {
  draw() {
    console.log(`Drawing Rectangle from ${this.start.x}:${this.start.y} to ${this.end.x}:${this.end.y}`);
  }

  clone(){
    return this
  }
}

const shapeList = [];
let shape = new Line(new Point(0, 0), new Point(10, 10));
shape.draw();
shapeList.push(shape);


shape = new Rectangle(new Point(1, 1), new Point(11,11))
shape.draw();
shapeList.push(shape)









